data/tutorials/wf_03_metaprogramming.md

```diff
@@ -0,0 +1,452 @@
+the rewritten parsetree the same way.
+
+But again, the tool responsible for driving the compilation being `dune`, using
+PPXes is just a matter of writing dune files. The same `preprocess` stanza
+should be used, this time with `pps`:
+
+```dune
+  (preprocess (pps ppx1 ppx2))
+```
+
+And that's all you need to use PPXes! Although these instructions will work for
+most PPXes, note that the first source of information will be the package
+documentation, as some PPXes might need some special care to integrate with
+dune.
+
+### Why are PPXes especially useful in OCaml
+
+Now that we know what is a PPX, and have seen example of such, let us see why it
+is particularly useful in OCaml.
+
+For one, the types are lost at execution time. That means that the structure of
+a type cannot be deconstructed at execution time to control the flow. This is
+why no general `to_string : 'a -> string` or `print : 'a -> ()` function can
+exist in compiled binaries (in the toplevel, types are kept).
+
+So any general function that depends on the structure of a type _must_ be
+written at compile time, when types are still available.
+
+For two, one of the strong features of OCaml is its type system, which can be
+used to check sanity of many things. An example of this is the [`Tyxml`
+library](https://github.com/ocsigen/tyxml), which uses the type system to ensure
+that the resulting HTML value satisfies most of the W3C standards. However,
+writing `Tyxml` values can be cumbersome compared to writing html syntax.
+Transforming HTML code into OCaml values at compile time allows to keep both the
+type-checker on the produced value and the familiarity of writing HTML code.
+
+Other rewriters such as `ppx_expect` show that being able to enrich the syntax
+via PPX rewriters is very useful, even outside of the specificity of OCaml.
+
+## The need for controlling the PPX ecosystem: Ppxlib
+
+Although PPXes are great for generating code at compile time, they raise a few
+questions, especially in the presence of multiple PPX rewriters:
+- What is the semantics of the composition of multiple PPXes? The order might
+  matter.
+- How can I trust that some part of my code are not modified, when using
+  third-party PPXes?
+- How can I keep short compilation time when using many rewriters?
+- How can I write a PPX easily if I have to deal with parsing or demarshalling
+  an AST, and deal with such a [long and complex
+  type](https://v2.ocaml.org/api/compilerlibref/Parsetree.html) as in parsetree?
+- How can I solve the problem that the OCaml syntax and the parsetree types
+  change with OCaml version?
+
+The good answer to these questions is a platform tool to write PPXes:
+[ppxlib](https://github.com/ocaml-ppx/ppxlib/). It takes care of easing the
+process of writing a PPX, ensure that composition is not too much of a problem,
+and factors as much as possible the work in case of multiple rewriters applied
+to a file.
```

Comment:
```suggestion
A good answer to these questions is to use a platform tool to write PPXs:
[ppxlib](https://github.com/ocaml-ppx/ppxlib/). It takes care of easing the
process of writing a PPX, ensures that composition is not too much of a problem,
and factors the work in case of multiple rewriters applied
to a file as much as possible.
```

Reply:
Here I think we want to be more explicit that it is the currently only supported option, as we want all ppx authors to use ppxlib. I'll try to reword that!